### PR TITLE
Improve support for runtime keepalive under node

### PR DESCRIPTION
**Pull request by sbc100**
Move `Browser.safeSetTimeout` to top level library function (since this  
code should work just as well under node).                                  
                                                                                                              
Test that `emscripten_async_call` works under node and that the runtime  
is kept alive until the callback fires.                                     
                                                                            
In order to allow for this I've improved the handling of EXIT_RUNTIME=0  
under node:                                                                 
                                                                            
By default (unless `EXIT_RUNTIME=1` is specified) emscripten programs       
running under node will no longer call `process.exit()` on `exit()`.        
Instead they will simply unwind the stack return to the event loop, much 
like they do on the web.  In many cases the node process will still exit 
naturally if there is nothing keeping the event loop going.  To             
guarantee that the node process exits immediately please build with         
`EXIT_RUNTIME=1`.

### CODE REVIEW

File: src/library_browser.js [hunk 1]
```diff
@@ -1421,13 +1421,14 @@ var LibraryBrowser = {
...
   emscripten_async_call__sig: 'viii',
+  emscripten_async_call__deps: ['$safeSetTimeout'],
   emscripten_async_call: function(func, arg, millis) {
     function wrapper() {
       {{{ makeDynCall('vi', 'func') }}}(arg);
     }
 
-    if (millis >= 0) {
-      Browser.safeSetTimeout(wrapper, millis);
+    if (millis >= 0 || ENVIRONMENT_IS_NODE) {
```
Comment (kripken) [inline src/library_browser.js:1430]: can guard with `#if ENVIRONMENT_MAY_BE_NODE`

File: src/library_browser.js [hunk 2]
```diff
@@ -488,20 +489,19 @@ var LibraryBrowser = {
 
     // abort and pause-aware versions TODO: build main loop on top of this?
 
+    safeSetTimeout: function(func) {
+#if ASSERTIONS
+      assert(false, "Please use `safeSetTimeout` instead");
```
Comment (kripken) [inline src/library_browser.js:494]:
```suggestion
      assert(false, "Please use `safeSetTimeout` instead of Browser.safeSetTimeout");
```

File: tests/test_other.py [hunk 3]
```diff
@@ -2602,9 +2602,9 @@ def test_demangle_cpp(self):
   def test_demangle_malloc_infinite_loop_crash(self):
     self.run_process([EMXX, test_file('malloc_demangle_infinite_loop.cpp'), '-g', '-s', 'ABORTING_MALLOC', '-s', 'DEMANGLE_SUPPORT'])
     output = self.run_js('a.out.js', assert_returncode=NON_ZERO)
-    if output.count('Cannot enlarge memory arrays') > 4:
+    if output.count('Cannot enlarge memory arrays') > 5:
       print(output)
-    self.assertLess(output.count('Cannot enlarge memory arrays'),  5)
+    self.assertLess(output.count('Cannot enlarge memory arrays'),  6)
```
Comment (kripken) [inline tests/test_other.py:2607]: what changed here?

File: tests/test_other.py [hunk 4]
```diff
@@ -9173,7 +9176,7 @@ def test_proxy_to_pthread_stack(self):
     # DEFAULT_PTHREAD_STACK_SIZE.
     self.do_smart_test(test_file('other/test_proxy_to_pthread_stack.c'),
                        ['success'],
-                       emcc_args=['-s', 'USE_PTHREADS', '-s', 'PROXY_TO_PTHREAD', '-s', 'DEFAULT_PTHREAD_STACK_SIZE=64kb', '-s', 'TOTAL_STACK=128kb'])
+                       emcc_args=['-s', 'USE_PTHREADS', '-s', 'PROXY_TO_PTHREAD', '-s', 'DEFAULT_PTHREAD_STACK_SIZE=64kb', '-s', 'TOTAL_STACK=128kb', '-s', 'EXIT_RUNTIME'])
```
Comment (kripken) [inline tests/test_other.py:9179]: why is this needed?

File: src/postamble.js [hunk 5]
```diff
@@ -474,7 +466,7 @@ function exit(status, implicit) {
     ABORT = true;
   }
 
-  quit_(status, new ExitStatus(status));
+  quit_(status, new ExitStatus(status), keepRuntimeAlive());
```
Comment (kripken) [inline src/postamble.js:469]: how about avoiding the `keepRuntimeAlive()` call here and on line 211, and instead doing it in the called function? That seems simpler too, as we always pass the same thing?

File: src/shell.js [hunk 6]
```diff
@@ -201,7 +201,11 @@ if (ENVIRONMENT_IS_NODE) {
   process['on']('unhandledRejection', abort);
 #endif
 
-  quit_ = function(status) {
+  quit_ = function(status, toThrow, keepalive) {
+    if (keepalive) {
```
Comment (kripken) [inline src/shell.js:205]: That is, I am proposing that this be
```suggestion
    if (keepRuntimeAlive()) {
```

File: src/library.js [hunk 7]
```diff
@@ -3691,6 +3690,14 @@ LibraryManager.library = {
   },
 #endif
 
+  $emSetTimeout: function(func, timeout) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/library.js]: Ah, looks like this was a comment I wrote that became stale by the time I finished the review!

File: src/shell.js (hunk 6 above)
Comment (sbc100) [reply src/shell.js:205]: Done

File: tests/test_other.py (hunk 4 above)
Comment (sbc100) [reply tests/test_other.py:9179]: This is a little complicated.

node was previously not honoring `EXIT_RUNTIME=0` (the default).   We were forcing an exit via `proces.exit` from `quit_` (itself called from from `exit`) even when we want to runtime to to stay alive.    With this change, the modified `quit_` will now to the right thing and just throw/unwind rather than `process.exit` giving the process a chance to stay alive after `exit`.

When callMain() catches the ExitStatus() is simply ignores it (its just signal to unwind and be done).   If the node event loop has no more work to do the entire process will then exit.. simply by coincidence not because emscripten asked for an explicit exit.   This means that the behaviour of `EXIT_RUNTIME=0` and `EXIT_RUNTIME=1` were (and continue to be) kind of the same under node for simple programs that are not doing other async work.   This is why we don't need `EXIT_RUNTIME=1` on *all* our node unit tests, because node will naturally exit in most cases.

The situation for threaded programs is a little different:  As soon as any thread is started (or added to the thread pool) I believe we prevent the node process from naturally exiting since it needs to stick around to talk communicate with the thread.  

This means that threaded programs under node will not naturally exit and now require `-s EXIT_RUNTIME=1` to exit at all.     Does this seems unreasonable?   Should we need another setting such as `-s PROCESS_EXIT_ON_EXIT` which defaults to true independently of `EXIT_RUNTIME`?

**Comment by sbc100** [comment]
I've added a ChangeLog entry that I think captures what has changed.   Let me know what you think.

File: tests/test_other.py (hunk 3 above)
Comment (sbc100) [reply tests/test_other.py:2607]: This is because the abort gets reported to stderr one more time by the node process because it is rethrown to the top level.  Previously we would call `process.exit` rather that rethrow to the outside node process.

This number is basically a record of amount times we repeat ourselves when an abort happens in JS :-/   We should probably address that.

File: tests/test_other.py (hunk 4 above)
Comment (kripken) [reply tests/test_other.py:9179]: That node needs `EXIT_RUNTIME` in pthreads builds doesn't seem too bad to me, if we document it in the changelog and maybe also pthreads.rst. Node pthreads is still somewhat experimental I'd say. I don't think we need a new setting.

File: ChangeLog.md [hunk 8]
```diff
@@ -20,6 +20,12 @@ See docs/process.md for more on how version tagging works.
 
 2.0.25
 ------
+- By default (unless `EXIT_RUNTIME=1` is specified) emscripten programs running
+  under node will no longer call `process.exit()` on `exit()`.  Instead they
+  will simply unwind the stack return to the event loop, much like they do
```
Comment (kripken) [inline ChangeLog.md:25]:
```suggestion
  will simply unwind the stack and return to the event loop, much like they do
```

File: ChangeLog.md [hunk 9]
```diff
@@ -20,6 +20,12 @@ See docs/process.md for more on how version tagging works.
...
+  on the web.  In many cases the node process will still exit naturally if
+  there is nothing keeping the event loop going.  To guarantee that the node
+  process exits immediately please build with `EXIT_RUNTIME=1`.
```
Comment (kripken) [inline ChangeLog.md:28]: What about error code reporting - is there some chance this causes a test to have a false positive or negative? That is, if it exits immediately, could it swallow an error? I think we have enough checks to avoid that, but I'm not 100% sure.

File: src/library_browser.js [hunk 10]
```diff
@@ -1421,13 +1421,19 @@ var LibraryBrowser = {
...
   emscripten_async_call__sig: 'viii',
+  emscripten_async_call__deps: ['$safeSetTimeout'],
   emscripten_async_call: function(func, arg, millis) {
     function wrapper() {
       {{{ makeDynCall('vi', 'func') }}}(arg);
     }
 
+#if ENVIRONMENT_MAY_BE_NODE
+    // node does not support requestAnimationFrame
+    if (millis >= 0 || ENVIRONMENT_IS_NODE) {
```
Comment (kripken) [inline src/library_browser.js:1432]: The `millis >= 0` check should not be in the ifdef I think.

File: src/preamble.js [hunk 11]
```diff
@@ -356,6 +356,11 @@ var __ATPOSTRUN__ = []; // functions called after the main() is called
 
 var runtimeInitialized = false;
 var runtimeExited = false;
+var runtimeKeepaliveCounter = 0;
+
+function keepRuntimeAlive() {
+  return noExitRuntime || runtimeKeepaliveCounter > 0;
+}
```
Comment (kripken) [inline src/preamble.js:363]: In general we prefer to put stuff in JS libraries, instead of always including it in preamble/postamble. Why does this do the opposite?

File: ChangeLog.md (hunk 9 above)
Comment (sbc100) [reply ChangeLog.md:28]: The code path in `quit_` calls `process['exitCode'] = status`.. so even if we exit "naturally" the correct return code will still be generated.

We have a fair bit of testing for this behavior (see `other.test_returncode`) so we know it works and we test a lot of difference circumstances there.

File: src/library_browser.js (hunk 10 above)
Comment (sbc100) [reply src/library_browser.js:1432]: I think I needs to be since `ENVIRONMENT_MAY_BE_NODE` is true by default isn't it?   If we want to use safeRequestAnimationFrame in the browser when `ENVIRONMENT_MAY_BE_NODE` is set don't we need it?

File: src/preamble.js (hunk 11 above)
Comment (sbc100) [reply src/preamble.js:363]: Since this is now being called from shell.js and basically always required I thought it is simpler in this particular case to add this tiny function to the preamble.js.    It also makes sense to me to have runtimeKeepaliveCounter  declared alongside the existing noExitRuntime since they work together.   Do you think I should revert this part of the change?

Comment (sbc100) [reply src/preamble.js:363]: The other reason for putting stuff in the library rather than preamble/postamble is so that it can be shared between MINIMAL_RUNTIME and the normal runtime (which have different preamble files), but MINIMAL_RUNTIME doesn't want this function so it makes less sense in the library (unlike for example `establishStackSpace` and `invokeEntryPoint` which are also force includes from the library).

Comment (kripken) [reply src/preamble.js:363]: I guess that's fair, yeah, as this is something not relevant for minimal runtime anyhow, and the code gets simpler, sgtm.

File: src/library_browser.js (hunk 10 above)
Comment (kripken) [reply src/library_browser.js:1432]: Oh, sorry, I see the millis part is duplicated, which confused me. I was looking for this:
```js
  if (millis >= 0
#ifdef NODE
      || NODE
#endif
    ) {
```

I lean towards that to reduce duplication, but I don't feel strongly.

**Review by kripken — APPROVED** [review]

File: src/library_browser.js (hunk 10 above)
Comment (sbc100) [reply src/library_browser.js:1432]: OK will do.   I'm a little allergic to using `#ifdef` to break up  single line like that, but maybe you are right.

**Comment by sbc100** [comment]
Updated ChangeLog to specifically mention the issue around multi-threaded programs

File: ChangeLog.md [hunk 12]
```diff
@@ -20,6 +20,14 @@ See docs/process.md for more on how version tagging works.
 
 2.0.25
 ------
+- By default (unless `EXIT_RUNTIME=1` is specified) emscripten programs running
+  under node will no longer call `process.exit()` on `exit()`.  Instead they
+  will simply unwind the stack and return to the event loop, much like they do
+  on the web.  In many cases the node process will then exit naturally if there
+  is nothing keeping the event loop going.  Note: Because of the way that
+  threads are implemented under node this means that multi-threaded programs now
+  requirte `EXIT_RUNTIME=1` in order to actually bring down the process on
```
Comment (kripken) [inline ChangeLog.md:29]:
```suggestion
  require `EXIT_RUNTIME=1` in order to actually bring down the process on
```

File: ChangeLog.md [hunk 13]
```diff
@@ -20,6 +20,14 @@ See docs/process.md for more on how version tagging works.
...
+  `exit()` (or return from main).
```
Comment (kripken) [inline ChangeLog.md:30]: Is the symptom here that a multithreaded node program will hang? I wonder if we can at least add a warning to help with that. Or at least to mention debugging tips here in the changelog.

Comment (kripken) [reply ChangeLog.md:30]: (That is, mention in the changelog that the symptom is hanging.)

Comment (sbc100) [reply ChangeLog.md:30]: It does not hang exactly, its just sits there with the event loop running waiting for more events to process.    Similar to if you did a `setInterval(1)` ..    This kind of makes sense since `EXIT_RUNTIME=0`.. in theory we could argue that is more correct to keep the event loop alive in this case.
